jupyterlab_nvdashboard/apps/gpu.py

```diff
@@ -201,48 +201,65 @@ def cb():
     doc.add_periodic_callback(cb, 200)
 
 
-def nvlink(doc):
-
-    import subprocess as sp
+def _get_nvlink_throughput():
+    throughput = [
+        pynvml.nvmlDeviceGetFieldValues(
+            handle,
+            [
+                (pynvml.NVML_FI_DEV_NVLINK_THROUGHPUT_DATA_RX, scope_id)
+                for scope_id in range(pynvml.NVML_NVLINK_MAX_LINKS)
+            ] +
+            [
+                (pynvml.NVML_FI_DEV_NVLINK_THROUGHPUT_DATA_RX, scope_id)
+                for scope_id in range(pynvml.NVML_NVLINK_MAX_LINKS)
+            ]
+        )
+        for handle in gpu_handles
+    ]
 
-    # Use device-0/link-0 to get "upper bound"
-    counter = 1
-    nlinks = pynvml.NVML_NVLINK_MAX_LINKS
-    nvlink_link_bw = {
-        # Keys = NVLink Version, Values = Max Link BW (per direction)
-        # [Note: Using specs at https://en.wikichip.org/wiki/nvidia/nvlink]
-        1: 20.0 * GB,  # GB/s
-        2: 25.0 * GB,  # GB/s
+    # Output is given in KiB, thus multiply by 1024 for result in bytes
+    # First `pynvml.NVML_NVLINK_MAX_LINKS` contain RX throughput, last
+    # `pynvml.NVML_NVLINK_MAX_LINKS` contain TX
+    return {
+        "rx": [
+            sum(t[i].value.ullVal * 1024
+                for i in range(pynvml.NVML_NVLINK_MAX_LINKS))
+            for t in throughput
+        ],
+        "tx": [
+            sum(t[pynvml.NVML_NVLINK_MAX_LINKS + i].value.ullVal * 1024
+                for i in range(pynvml.NVML_NVLINK_MAX_LINKS))
+            for t in throughput
+        ],
     }
-    # Max NVLink Throughput = BW-per-link * nlinks
-    max_bw = nlinks * nvlink_link_bw.get(nvlink_ver, 25.0 * GB)
-
-    # nvmlDeviceSetNvLinkUtilizationControl seems limited, using smi:
-    sp.call(
-        [
-            "nvidia-smi",
-            "nvlink",
-            "--setcontrol",
-            str(counter) + "bz",  # Get output in bytes
-        ]
-    )
+
+
+def _get_max_bandwidth():
+    links = [
+        getattr(pynvml, f"NVML_FI_DEV_NVLINK_SPEED_MBPS_L{i}")
```

Comment:
Should we pin the oldest version of pynvml to 11 in `requirements.txt`? (I don't think these variables existed before 11.0.0)

Reply:
Yes, this is a good idea. Done in https://github.com/rapidsai/jupyterlab-nvdashboard/pull/133/commits/3783d43fb8c4182771e0e27ecd13a0b69d45882c